### PR TITLE
fix: Bump Abseil to fix Linux build issues

### PR DESCRIPTION
**Pull request by martijneken**
Pick up this fix: https://github.com/abseil/abseil-cpp/pull/1187

Bump to latest LTS version (past Envoy) to pick up another fix found in local fuzz tests:
https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/399#issuecomment-2261871275

Build issue seen on:
```
$ /usr/lib/llvm-16/bin/clang --version
Debian clang version 16.0.6 (26)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-16/bin
```

### CODE REVIEW

File: bazel/repositories.bzl [hunk 1]
```diff
@@ -82,7 +82,18 @@ def proxy_wasm_cpp_host_repositories():
         patch_args = ["-p1"],
     )
 
-    # Core.
+    # Core deps. Try to keep up with Envoy versions:
+    # https://github.com/envoyproxy/envoy/blob/main/bazel/repository_locations.bzl
```
Comment (PiotrSikora) [inline bazel/repositories.bzl:86]: Why does it matter whether C++ Host version match Envoy's? The builds in Envoy don't rely on versions in here, and the builds in here don't rely on anything in Envoy.

A lot of dependencies were not updates in the past 2 years, so things broke, but virtually all of those issues are due to the changes in compilers and build rules/systems, and have nothing to do with Envoy.

File: bazel/repositories.bzl [hunk 2]
```diff
@@ -82,7 +82,18 @@ def proxy_wasm_cpp_host_repositories():
...
+
+    # Note: latest LTS release, ahead of Envoy to pick up a bugfix found in local fuzzing:
+    # https://github.com/abseil/abseil-cpp/commit/e7858c73279d81cbc005d9c76a385ab535520635
+    maybe(
+        http_archive,
+        name = "com_google_absl",
+        sha256 = "733726b8c3a6d39a4120d7e45ea8b41a434cdacde401cba500f14236c49b39dc",
+        strip_prefix = "abseil-cpp-20240116.2",
+        urls = ["https://github.com/abseil/abseil-cpp/archive/20240116.2.tar.gz"],
+    )
```
Comment (PiotrSikora) [inline bazel/repositories.bzl:96]: A few things:
1) Again, this is unrelated to Envoy.
2) C++ Host does not rely on abseil, so it's weird adding it as a dependency here.
3) I believe this should be fixed by the update of `rules_fuzzing` to a newer version (which ships with abseil v20240116.1).

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

**Comment by PiotrSikora** [comment]
Also, does the issue show up only in your local testing? For what invocation?

I'm not seeing any failures either on the CI or locally without this PR on Debian with GCC or Clang 14-19, so perhaps there is some weird issue related to your setup?

File: bazel/repositories.bzl (hunk 1 above)
Comment (martijneken) [reply bazel/repositories.bzl:86]: Errr it doesn't have to match exactly, but we also can't lag dependencies by years (the current state), given that part of the CppHost verification process is to patch a draft PR into Envoy and ensure everything works. Version mismatches are a drag on productivity in this repo. I don't know of a systematic way to address this since Envoy likes to be in control of its dependencies.

File: bazel/repositories.bzl (hunk 2 above)
Comment (martijneken) [reply bazel/repositories.bzl:96]: > update of rules_fuzzing to a newer version

Thanks for identifying the Abseil dependency path. If only it were this simple. The next rules_fuzzing major version (v0.4) adds some dynamically loaded pydeps that I can't figure out (they want to be loaded via WORKSPACE...), and the latest version (v0.5) also bumps to Bazel 7.

Can we do the tactical (Abseil) fix while we sort out larger and more intrusive upgrades?

File: bazel/repositories.bzl (hunk 1 above)
Comment (PiotrSikora) [reply bazel/repositories.bzl:86]: > Errr it doesn't have to match exactly, but we also can't lag dependencies by years (the current state)

Right, but that can be shortened to "Keep them updated." ;)

> given that part of the CppHost verification process is to patch a draft PR into Envoy and ensure everything works.

That should require only bumping C++ Host, and sporadically a Wasm engine dependency, if that's what's updated behind the scenes.

There is a separate issue with Wasmtime, but that's because we've embedded Wasmtime's Bazel rules into this repository. I've suggested splitting it into `wasmtime-bazel` multiple times in the past, since it really shouldn't be here.

File: bazel/repositories.bzl (hunk 2 above)
Comment (PiotrSikora) [reply bazel/repositories.bzl:96]: If updating `rules_fuzzing` is problematic, then I guess we have to override abseil dependency here.

Could you change/update the comment to reflect that this is temporary override and that it should be removed once we update `rules_fuzzing`? Otherwise, it's going to stay here as an explicit dependency forever.

Perhaps a file an issue for that as well?

**Comment by PiotrSikora** [comment]
> Build issue seen on:
> ```
> $ /usr/lib/llvm-16/bin/clang --version
> Debian clang version 16.0.6 (26)
> Target: x86_64-pc-linux-gnu
> Thread model: posix
> InstalledDir: /usr/lib/llvm-16/bin
> ```

That's interesting, because it works fine on:
```
$ clang-16 --version
Debian clang version 16.0.6 (15~deb12u1)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

What's the invocation that's broken?

Note that [abseil/abseil-cpp#1187](https://github.com/abseil/abseil-cpp/pull/1187) talks about GCC 13 and not Clang.

File: bazel/repositories.bzl (hunk 1 above)
Comment (martijneken) [reply bazel/repositories.bzl]: > shortened to "Keep them updated." ;)

Fair, done.

**Comment by martijneken** [comment]
> What's the invocation that's broken?

```
$ bazelisk test --verbose_failures --define engine=null --config=clang -- //test/...
...
ERROR: /usr/local/google/home/mstevenson/.cache/bazel/_bazel_mstevenson/35e00757f2b858287fbaf6f5989ae0e8/external/com_google_absl/absl/strings/BUILD.bazel:1099:11: Compiling absl/strings/internal/str_format/float_conversion.cc fai
led: (Exit 1): clang failed: error executing command                                                                                                                                                                                  
  (cd /usr/local/google/home/mstevenson/.cache/bazel/_bazel_mstevenson/35e00757f2b858287fbaf6f5989ae0e8/sandbox/linux-sandbox/237/execroot/proxy_wasm_cpp_host && \                                                                   
  exec env - \                                                                                                                                                                                                                        
    BAZEL_COMPILER=clang \                                                                                                                                                                                                            
    CC=clang \                                                                                                                                                                                                                        
    CXX=clang++ \                                                                                                                                                                                                                     
    PATH=<...> \                        
    PWD=/proc/self/cwd \                                                                                                                                                                                                              
  /usr/lib/llvm-16/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/k8
-fastbuild-ST-5c2a6e57c052/bin/external/com_google_absl/absl/strings/_objs/str_format_internal/float_conversion.pic.d '-frandom-seed=bazel-out/k8-fastbuild-ST-5c2a6e57c052/bin/external/com_google_absl/absl/strings/_objs/str_format
_internal/float_conversion.pic.o' -fPIC -iquote external/com_google_absl -iquote bazel-out/k8-fastbuild-ST-5c2a6e57c052/bin/external/com_google_absl -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION '-std=c++17' -Wall -Wextra -Wcast-qual
 -Wconversion -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Winfinite-recursion -Winvalid-constexpr -Wliteral-conversion -Wmissing-declarations -Woverlength-string
s -Wpointer-arith -Wself-assign -Wshadow-all -Wstring-conversion -Wtautological-overlap-compare -Wundef -Wuninitialized -Wunreachable-code -Wunused-comparison -Wunused-local-typedefs -Wunused-result -Wvla -Wwrite-strings -Wno-floa
t-conversion -Wno-implicit-float-conversion -Wno-implicit-int-float-conversion -Wno-implicit-int-conversion -Wno-shorten-64-to-32 -Wno-sign-conversion -Wno-unknown-warning-option -DNOMINMAX -no-canonical-prefixes -Wno-builtin-macr
o-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_absl/absl/strings/internal/str_format/float_conversion.cc -o bazel-out/k8-fastbuild-ST-5c2a6e57c052/bin/external/com_g
oogle_absl/absl/strings/_objs/str_format_internal/float_conversion.pic.o)                                                                                                                                                             
# Configuration: f5c41261913aa58b27c39d480bafcec8c7dcbd34b432916aa8bb476b1c64edba                                                                                                                                                     
# Execution platform: @local_config_platform//:host                                                                                                                                                                                   
...
In file included from external/com_google_absl/absl/strings/internal/str_format/float_conversion.cc:15:
In file included from external/com_google_absl/absl/strings/internal/str_format/float_conversion.h:18:
external/com_google_absl/absl/strings/internal/str_format/extension.h:34:35: error: unknown type name 'uint8_t'
enum class FormatConversionChar : uint8_t;
                                  ^
external/com_google_absl/absl/strings/internal/str_format/extension.h:35:38: error: unknown type name 'uint64_t'
enum class FormatConversionCharSet : uint64_t;
                                     ^
external/com_google_absl/absl/strings/internal/str_format/extension.h:131:20: error: unknown type name 'uint8_t'
enum class Flags : uint8_t {
                   ^
```

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Comment by PiotrSikora** [comment]
> ```
> $ bazelisk test --verbose_failures --define engine=null --config=clang -- //test/...
> [...]
> external/com_google_absl/absl/strings/internal/str_format/extension.h:131:20: error: unknown type name 'uint8_t'
> enum class Flags : uint8_t {
>                    ^
> ```

FWIW, it works here.

**Comment by martijneken** [comment]
> FWIW, it works here.

Any interest in pursuing hermetic Bazel toolchains? (toolchains_llvm, hermetic_cc_toolchain)

**Comment by PiotrSikora** [comment]
> Any interest in pursuing hermetic Bazel toolchains? (toolchains_llvm, hermetic_cc_toolchain)

Sure, I believe that @leonm1 signed up for that a while ago.

**Review by mpwarres — APPROVED** [review]
